sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuCast.scala

```diff
@@ -510,126 +510,71 @@ case class GpuCast(
-    inputSchema: Array[StructField]): ColumnVector = {
-    var separatorColumn: ColumnVector = null
-    var spaceColumn: ColumnVector = null
-    val columns: ArrayBuffer[ColumnVector] = new ArrayBuffer[ColumnVector]()
 
-    try {
-      withResource(GpuScalar.from(", ", StringType)) { separatorScalar =>
-        separatorColumn = ColumnVector.fromScalar(separatorScalar, input.getRowCount.toInt)
-      }
-      withResource(GpuScalar.from("{", StringType)) { bracketScalar =>
-        columns += ColumnVector.fromScalar(bracketScalar, input.getRowCount.toInt)
-      }
-
-      withResource(input.getChildColumnView(0)) { childView =>
-        columns += doColumnar(childView, inputSchema(0).dataType)
-      }
-      for(childIndex <- 1 until input.getNumChildren()) {
-        withResource(input.getChildColumnView(childIndex)) { childView =>
-          columns += separatorColumn
-          columns += doColumnar(childView, inputSchema(childIndex).dataType)
+        columns += rightColumn.incRefCount()
+        withResource(
+          ColumnVector.stringConcatenate(emptyScalar, nullScalar, columns.toArray)
+        ) { res =>
+          // null rows
+          res.mergeAndSetValidity(BinaryOp.BITWISE_AND, input)
         }
-      }
-      withResource(GpuScalar.from("}", StringType)) { bracketScalar =>
-        columns += ColumnVector.fromScalar(bracketScalar, input.getRowCount.toInt)
-      }
-
-      // Merge casted child columns
-      withResource(GpuScalar.from("", StringType)) { emptyStrScalar =>
-        withResource(GpuScalar.from("null", StringType)) { nullStringScalar =>
-          withResource(ColumnVector.stringConcatenate(emptyStrScalar, nullStringScalar,
-            columns.toArray[ColumnView])) { fullResult =>
-            // Reflect the struct column's validity vector in the result
-            fullResult.mergeAndSetValidity(BinaryOp.BITWISE_AND, input)
-          }
-        }
-      }
-    } finally {
-      if (separatorColumn != null) {
-        columns.foreach(col =>
-          if(col.getNativeView() != separatorColumn.getNativeView()) {
-            col.close()
-          })
-        separatorColumn.close()
-      }
-      if (spaceColumn != null) {
-        spaceColumn.close()
+      } finally {
+        columns.foreach(_.close())
       }
     }
-  }
 
-  private def castStructToString(input: ColumnView,
-    legacyCastToString: Boolean, inputSchema: Array[StructField]): ColumnVector = {
-    if (legacyCastToString) {
-      legacyStructToString(input, inputSchema)
-    } else {
-      modernStructToString(input,inputSchema)
+    withResource(Array(
+      Scalar.fromString(emptyStr),
+      Scalar.fromString(nullStr),
+      Scalar.fromString(separatorStr),
+      Scalar.fromString(spaceStr),
+      Scalar.fromString(leftStr),
+      Scalar.fromString(rightStr)
+    )) { case Array(emptyScalar, nullScalar, columnScalars@_*) =>
+      withResource(columnScalars.map(s => ColumnVector.fromScalar(s, numRows)))(columns => {
```

Comment:
If the `map` throws while trying to build the sequence then we can leak
```suggestion
      withResource(columnScalars.safeMap(s => ColumnVector.fromScalar(s, numRows)))(columns => {
```